### PR TITLE
Prepare release 2.4.0

### PR DESCRIPTION
**Pull request by hramezani**
TODO:
* [x] #7569
* [x] #7567 
* [x] #6820
* [x] #7417
* [x] #7589
* [x] Once #7417 is merged I should double check all paths on the current sites are redirected properly

Todo after release:
* [ ] https://github.com/samuelcolvin/mkdocs-run-code/pull/1

### CODE REVIEW

**Comment by samuelcolvin** [comment]
It looks like the heading levels are wrong, I think we should remove "What Changed".

@hramezani please update the the change log.

File: HISTORY.md [hunk 1]
```diff
@@ -1,3 +1,70 @@
+## v2.4.0 (2023-09-22)
+
+[GitHub release](https://github.com/pydantic/pydantic/releases/tag/v2.4.0)
+
+## What's Changed
```
Comment (samuelcolvin) [inline HISTORY.md:5]: Remove this

Comment (hramezani) [reply HISTORY.md]: done

File: HISTORY.md [hunk 2]
```diff
@@ -1,3 +1,69 @@
...
+* Standardize `__get_pydantic_core_schema__` signature by @hramezani in [#7415](https://github.com/pydantic/pydantic/pull/7415)
+* Fix Generic Dataclass Fields Mutation Bug (when using `TypeAdapter`) by @sydney-runkle in [#7435](https://github.com/pydantic/pydantic/pull/7435)
+* Fix `TypeError` on `model_validator` in `wrap` mode by @pmmmwh in [#7496](https://github.com/pydantic/pydantic/pull/7496)
+* Improve enum error message by @hramezani in [#7506](https://github.com/pydantic/pydantic/pull/7506)
+* Make `repr` work for instances that failed initialization when handling `ValidationError`s by @dmontagu in [#7439](https://github.com/pydantic/pydantic/pull/7439)
+* Fixed a regular expression denial of service issue by limiting whitespaces by @prodigysml in [#7360](https://github.com/pydantic/pydantic/pull/7360)
+* Fix handling of `UUID` values having `UUID.version=None` by @lig in [#7566](https://github.com/pydantic/pydantic/pull/7566)
+* Fix `__iter__` returning private `cached_property` info by @sydney-runkle in [#7570](https://github.com/pydantic/pydantic/pull/7570)
+
+## New Contributors
```
Comment (davidhewitt) [inline HISTORY.md:54]:
```suggestion
### New Contributors
```

File: HISTORY.md [hunk 3]
```diff
@@ -1,3 +1,69 @@
...
+* Add `BaseModel.model_validate_strings` and `TypeAdapter.validate_strings` by @hramezani in [#7552](https://github.com/pydantic/pydantic/pull/7552)
+* Add Pydantic `plugins` experimental implementation by @lig @samuelcolvin and in [#6820](https://github.com/pydantic/pydantic/pull/6820)
+
+### Changes
+
+* Do not override `model_post_init` in subclass with private attrs by @Viicos in [#7302](https://github.com/pydantic/pydantic/pull/7302)
+* Make fields with defaults not required in the serialization schema by default by @dmontagu in [#7275](https://github.com/pydantic/pydantic/pull/7275)
+* Mark `Extra` as deprecated by @disrupted in [#7299](https://github.com/pydantic/pydantic/pull/7299)
+* Make `EncodedStr` a dataclass by @Kludex in [#7396](https://github.com/pydantic/pydantic/pull/7396)
+* Move `annoated_handlers` to be public by @samuelcolvin in [#7569](https://github.com/pydantic/pydantic/pull/7569)
```
Comment (davidhewitt) [inline HISTORY.md:23]:
```suggestion
* Move `annotated_handlers` to be public by @samuelcolvin in [#7569](https://github.com/pydantic/pydantic/pull/7569)
```

File: HISTORY.md [hunk 4]
```diff
@@ -1,3 +1,69 @@
...
+
+### Performance
+
+* Simplify flatteining and inlining of Coreschema by @adriangb in [#7523](https://github.com/pydantic/pydantic/pull/7523)
```
Comment (davidhewitt) [inline HISTORY.md:27]:
```suggestion
* Simplify flattening and inlining of `CoreSchema` by @adriangb in [#7523](https://github.com/pydantic/pydantic/pull/7523)
```

File: HISTORY.md [hunk 5]
```diff
@@ -1,3 +1,69 @@
...
+* Remove unused copies in CoreSchema walking by @adriangb in [#7528](https://github.com/pydantic/pydantic/pull/7528)
```
Comment (davidhewitt) [inline HISTORY.md:28]:
```suggestion
* Remove unused copies in `CoreSchema` walking by @adriangb in [#7528](https://github.com/pydantic/pydantic/pull/7528)
```

**Review by samuelcolvin — APPROVED** [review]

**Pull request merged**